rfcs/convergence/no-css-shorthands-in-make-styles.md

```diff
@@ -0,0 +1,218 @@
+
+The motivation is simple: no one sees console noise, in big applications it usually 100+ development warnings only on initial load.
+
+### Keep CSS properties
+
+- Keep interop with any CSS var on any CSS property
+- Keep all style properties to spec with CSS properties
+
+This means that custom syntax is a **discarded solution**, for example:
+
+```ts
+// ⚠ discarded solution, used only for an example
+makeStyles({
+  rootA: {
+    // 1️⃣ no custom properties
+    paddingX: 'var(--foo)',
+  },
+  rootB: {
+    // 2️⃣ no custom values
+    padding: { left: 'var(--foo)', right: 'var(--foo)' },
+    margin: { all: 'var(--foo)' },
+  },
+  rootC: {
+    // 2️⃣ no custom values
+    padding: ['var(--foo)', 'var(--foo)'],
+  },
+});
+```
+
+## Detailed Design or Proposal
+
+### Ban CSS shorthands in typings
+
+The first part of proposal is to ban CSS shorthands in typings: no shorthand properties = no need to expand them = no issues ✅
+
+Proposed to forbid usage of all CSS shorthands ([list of CSS shorthands exported from MDN](https://csstree.github.io/docs/syntax/#report&noedit&title=CSS%20shorthand%20properties%20by%20MDN%20data&q=JGlzQXJyYXk6ID0%2BICQgIT0gKCQgKyAnJyk7Cm1kbi5wcm9wZXJ0aWVzLmVudHJpZXMoKS4oeyBpZHgsIG5hbWU6IGtleSwgLi4uKHZhbHVlIHwgeyBjb21wdXRlZCwgc3ludGF4IH0pIH0pLltjb21wdXRlZC4kaXNBcnJheSgpXQ%3D%3D&v=ewogICAgdmlldzogJ3RhYmxlJywKICAgIGNvbHM6IHsKICAgICAgICBpZHg6ICd0ZXh0OiMuaW5kZXggKyAxJywKICAgICAgICBjb21wdXRlZDogewogICAgICAgICAgICBoZWFkZXI6ICdwcm9wcycsCiAgICAgICAgICAgIGNvbnRlbnQ6ICd1bDpjb21wdXRlZCcKICAgICAgICB9CiAgICB9LAogICAgbGltaXQ6IGZhbHNlCn0%3D)) except vendor prefixed (i.e. `-ms`, `-webkit`).
+
+```js
+makeStyles({
+  // 🚨 TypeScript will throw
+  rootA: { padding: '4px' },
+  // ✅ TypeScript will pass
+  rootB: { paddingLeft: '4px', paddingRight: '4px', paddingBottom: '4px', paddingTop: '4px' },
+});
+```
+
+### Create macro functions for expansion
+
+As it will be annoying and redundant to use CSS longhands in every case we will have special functions for expansion.
+
+```js
+import { macros } from '@fluentui/react-make-styles';
+
+makeStyles({
+  rootA: { paddingLeft: '4px', paddingRight: '4px', paddingBottom: '4px', paddingTop: '4px' },
+  rootB: {
+    ...macros.padding('4px'),
+    ...macros.padding('4px', '8px'),
+    ...macros.padding('4px', '8px', '16px', 0),
+  },
+});
+```
+
+`macros.padding('4px')` from the example above will expand it two four properties:
+
+```js
+shallowEqual(
+  { paddingLeft: '4px', paddingRight: '4px', paddingBottom: '4px', paddingTop: '4px' },
+  macros.padding('4px'),
+); // returns "true"
+```
+
+Each macro will implement arguments order & expand based on CSS spec.
+
+For initial implementation following macros will be implemented:
+
+- `border`
+- `borderLeft`
+- `borderBottom`
+- `borderRight`
+- `borderTop`
+- `borderColor`
+- `borderStyle`
+- `borderRadius`
+- `borderWidth`
+- `margin`
+- `padding`
```

Comment:
Consider adding background.
Lower priority, it would be a bonus to have a few for common shorthand in flexbox and grid.

Reply:
Yeah it's one of the first things that I was thinking, but it's not obvious unfortunately as `background` value could also image:

```tsx
// ✅ Easy-peasy
shorthands.background("green");
// ➡️ { backgroundColor: 'green' }

// 💥 Oops
shorthands.background('url("lizard.png")');
// ➡️ { backgroundColor: 'url("lizard.png")' }
```

https://developer.mozilla.org/en-US/docs/Web/CSS/background